### PR TITLE
Initial draft CICD Connector tutorial

### PR DESCRIPTION
**Pull request by janstaelensskyline**
Initial review can already be done.
The parts about the Validator still have TODO. These will be added in the next few days.

Do not merge yet.

### CODE REVIEW

**Comment by MariekeGO** [comment]
I've set the PR to draft for now, since it can't be merged yet.

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+1. Add 3 new steps under the current steps that:
+
+    1. Installs the Validators tool by adding a new job
+
+    1. Run the validator on your connector
+
+    1. Displays the results and optionally performs quality gating activities
+
+```yml
+    - name: InstallConnectorValidator
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:128]: Split up the word in multiple words. There is no need to have no spaces.

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+## Step 4: Extend the workflow with CD, automatic deployment
+
+1. Navigate to Actions
+
+1. On the left, click on your *.NET* workflow
+
+1. On the top, click on *dotnet.yml*. On the new window, click on the pencil icon to edit the page.
+
+1. Add a new Job called CD running on a windows image with 2 new steps that:
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:178]: Job can probably be lowercase?

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+      TODO
+      run: dataminer-connector-validate ${{ github.workspace }} --name HelloFromGithubWindows --output ${{ github.workspace }} --type automation
+
+    - name: QualityGate
+      TODO 
+```
+
+1. Commit your changes
+
+1. Navigate to Actions and see the run of your enhanced pipeline. This will perform:
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:162]: Add * around the Actions (consistency with previous Actions)

Comment (MichielOda) [reply develop/CICD/Tutorials/CICD_Tutorial_Connector.md:162]: Check later in document as well

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+    1. DataMiner Connector Validator
+
+    1. Package Creation
+
+    1. Direct Agent Deployment
+
+## Step 4: Add your GitHub secrets
+
+1. Navigate to Settings
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:232]: Same remarks as for Actions, add the * around it

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+1. Navigate to Actions
+
+1. Select the *.NET* workflow on the left
+
+1. Click on *Run workflow*
+
+1. You should now see your CI and CD jobs complete successfully
+
+1. Taking a screenshot of this successful run and sending that to thunder@skyline.be or uploading it through the [DoJo tutorials page](https://community.dataminer.services/learning-courses-tutorials/) will grant you DevOps points.
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:254]: DoJo => Dojo

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+## Advanced Options: CI
+
+Skyline Communications organization uses more than these actions. For enhanced CI, You can take a look at our [reusable workflow](https://github.com/SkylineCommunications/_ReusableWorkflows/blob/main/.github/workflows/Connector%20Master%20SDK%20Workflow.yml):
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:258]: You => you

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+1. SonarCloud static analysis to our CI
+
+1. Ability to use our GitHub Organization private NuGet store
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:262]: This is only applicable when they have been added as external collaborator to the repository of that NuGet package.

**Review by MichielOda — CHANGES_REQUESTED** [review]

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+```yml
+    - name: Install Connector Validator
+      run: dotnet tool install -g Skyline.DataMiner.CICD.Tools.Validator
+
+    - name: Run Connector Validator
+      run: dataminer-validator validate-protocol-solution --solution-path "${{ github.workspace }}" --output-directory "${{ github.workspace }}" --output-file-name "validateResults" 
+
+    - name: Archive Results
+      if: success()
+      uses: actions/upload-artifact@v2
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:136]: v4 already exists

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+    1. Compilation
+
+    1. Unit Testing
+
+    1. DataMiner Connector Validator
+
+## Step 4: Extend the workflow with CD, automatic deployment
+
+1. Navigate to Actions
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:172]: Add * around Actions

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+>[!IMPORTANT]
+> Deployment with a local artifact requires DataMiner 10.3.0/10.3.2 or higher
+
+```yml
+  CD:
+
+    runs-on: windows-latest
+    needs: CI
+    steps:
+    - uses: actions/checkout@v3
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:202]: v4 is already out

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+- [Microsoft Visual Studio](https://visualstudio.microsoft.com/downloads/)
+
+- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+
+- [GitHub account](https://docs.github.com/en/get-started/signing-up-for-github/signing-up-for-a-new-github-account)
+
+- [DataMiner Integration Studio](https://community.dataminer.services/exphub-dis/)
+
+> [!IMPORTANT]
+> The Validator CI step used in this tutorial, provided by Skyline DataMiner, only works for SDK-style projects targeting **.NET 4.8**.
```
Comment (MichielOda) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:29]: Targeting .NET 4.8 requirement is not needed anymore

**Review by MichielOda — CHANGES_REQUESTED** [review]

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+
+Expected duration: 20 minutes.
+
+## Prerequisites
+
+- Runtime .NET SDK 8.0. ([download](https://dotnet.microsoft.com/en-us/download/dotnet/8.0))
+
+- A staging DataMiner Agent that is accessible from your pipeline and that uses DataMiner version 10.3.0/10.3.2 or higher
+
+- [Microsoft Visual Studio](https://visualstudio.microsoft.com/downloads/)
```
Comment (MariekeGO) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:22]: Is there a minimum version of Visual Studio for this tutorial?

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+1. Select the *.NET* workflow on the left.
+
+1. Click *Run workflow*.
+
+You should now see your CI and CD jobs complete successfully.
+
+> [!NOTE]
+> Taking a screenshot of this successful run and sending that to <thunder@skyline.be> or uploading it through the [Dojo tutorials page](https://community.dataminer.services/learning-courses-tutorials/) will grant you DevOps points.
+
+## Advanced Options: CI
```
Comment (MariekeGO) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:278]: I'm not sure if these last two sections (from this line onwards) really belong in this tutorial, as they don't include something the user should do as part of the tutorial. It might be better to include this on a different page and then add a link to that page here.

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md (hunk 12 above)
Comment (spetebroodt) [reply develop/CICD/Tutorials/CICD_Tutorial_Connector.md:22]: After discussing it with @janstaelensskyline , it turns out this procedure is for VS 2022.

Comment (MariekeGO) [reply develop/CICD/Tutorials/CICD_Tutorial_Connector.md:22]: Could you add this to this prerequisite?

**Comment by spetebroodt** [comment]
@janstaelensskyline , I've tested the entire procedure using VS2022 and added some more detail to the different steps.
Can you have another look at the entire text to check whether everything is correct? There's also still an open question from Marieke.

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md (hunk 13 above)
Comment (janstaelensskyline) [reply develop/CICD/Tutorials/CICD_Tutorial_Connector.md]: This is mentioned during the Kata as well, I'd prefer to keep it as it links to more advanced information directly from where they just did the tutorial. The information is available under CI/CD docs as well I think. Though not part of the tutorial, it gives a "what's next?" for the users.

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 14]
```diff
@@ -197,102 +198,94 @@ Expected duration: 20 minutes.
...
+1. Go to the *Settings* tab.
 
-    1. SERVER_LOCATION
+1. On the left, click *Secrets and variables*, and then click *Actions*.
 
-    1. DATAMINER_USER
+1. Add the following secrets (by clicking *New repository secret*, entering the name of the secret, and clicking *Add secret*):
 
-    1. DATAMINER_PASSWORD
+   1. SERVER_LOCATION
```
Comment (janstaelensskyline) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:254]: Enters here?

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 15]
```diff
@@ -134,60 +137,58 @@ Expected duration: 20 minutes.
...
-   This will perform:
-
-    1. Compilation
+1. Go to the *Actions* tab, and check the run of your enhanced pipeline.
 
-    1. Unit Testing
+   The following will be performed:
 
-    1. DataMiner Connector Validator
+   1. Compilation
```
Comment (janstaelensskyline) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:185]: No enters needed?

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 16]
```diff
@@ -197,102 +198,94 @@ Expected duration: 20 minutes.
...
 
-    1. Unit testing
+   The following will be performed:
 
-    1. DataMiner connector validation
-
-    1. Package creation
-
-    1. Direct Agent deployment
+   1. Compilation
```
Comment (janstaelensskyline) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:240]: Are there enters needed?

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 17]
```diff
@@ -197,102 +198,94 @@ Expected duration: 20 minutes.
...
 1. Click *Run workflow*.
 
 You should now see your CI and CD jobs complete successfully.
 
 > [!NOTE]
-> Taking a screenshot of this successful run and sending that to <thunder@skyline.be> or uploading it through the [Dojo tutorials page](https://community.dataminer.services/learning-courses-tutorials/) will grant you DevOps points.
+> To be granted DevOps points, take a screenshot of this successful run and either send it to <thunder@skyline.be> or upload it via the [Dojo tutorials page](https://community.dataminer.services/learning-courses-tutorials/).
 
-## Advanced Options: CI
+## Advanced options: CI
```
Comment (janstaelensskyline) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:271]:
```suggestion
## What's next?
### Advanced options: CI
```

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 18]
```diff
@@ -197,102 +198,94 @@ Expected duration: 20 minutes.
...
+1. Tagging is considered a release cycle. This overrides the dmprotocol version with the tag.
 
-1. A regular commit and push is considered a build cycle, this adds a _Bx to the dmprotocol version. With x being the run-number of the pipeline.
+1. A regular commit and push is considered a build cycle. This adds a *_Bx* suffix to the dmprotocol version (*x* being the run number of the pipeline).
 
-1. We upload the dmprotocol as a downloadable artifact in GitHub
+1. We upload the dmprotocol as a downloadable artifact in GitHub.
 
-## Advanced Options: CD
+## Advanced options: CD
```
Comment (janstaelensskyline) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:285]:
```suggestion
### Advanced options: CD
```

File: develop/CICD/Tutorials/CICD_Tutorial_Connector.md [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+
+1. SonarCloud static analysis to our CI.
+
+1. Ability to use our GitHub Organization private NuGet store (when added as external collaborator to the package source repository).
+
+1. Tagging is considered a release cycle. This overrides the dmprotocol version with the tag.
+
+1. A regular commit and push is considered a build cycle. This adds a *_Bx* suffix to the dmprotocol version (*x* being the run number of the pipeline).
+
+1. The dmprotocol is uploaded as a downloadable artifact in GitHub.
```
Comment (MariekeGO) [inline develop/CICD/Tutorials/CICD_Tutorial_Connector.md:292]: @janstaelensskyline 
These list items seem to be mix of information and things that happen in the workflow. Given the numbered list format, I'd expect each list item to be a step in the workflow. If this should indeed be the case, could you rephrase the steps so that each item indeed describes an action that happens in the workflow? Or if that is not what you meant, could you use a bulleted list instead and rephrase the lines without verb (i.e. the first 2) so that they become full sentences like the rest of the list?

**Review by MariekeGO — APPROVED** [review]
